src/compile.rs

```diff
@@ -479,6 +478,145 @@ fn validate_module_contents<'a>(
     ret
 }
 
+pub fn validate_systems<'a, 'b>(
+    policies: &'b [PolicyFile],
+    module_map: &'b ModuleMap,
+    system_map: &'a mut SystemMap<'b>,
+) -> Result<(), CascadeErrors> {
+    let mut errors = CascadeErrors::new();
+
+    // Store all systems across files in a vector
+    let mut systems_vec: Vec<(&SimpleFile<String, String>, &System)> = Vec::new();
+    for p in policies {
+        for e in &p.policy.exprs {
+            if let Expression::Decl(Declaration::System(s)) = e {
+                systems_vec.push((&p.file, s));
+            }
+        }
+    }
+
+    for (file, system) in &systems_vec {
+        let mut system_modules = BTreeSet::new();
+        let mut configs = BTreeMap::new();
+
+        // Validate that the modules of a system exist
+        for m in &system.modules {
+            match module_map.get(m.as_ref()) {
+                Some(module) => {
+                    system_modules.insert(module);
+                }
+                None => errors.append(CascadeErrors::from(
+                    ErrorItem::make_compile_or_internal_error(
+                        &format!("Module {} does not exist", m.as_ref()),
+                        Some(file),
+                        m.get_range(),
+                        "modules within systems must be declared elsewhere",
+                    ),
+                )),
+            }
+        }
+        // Validate the system's configurations
+        for c in &system.configurations {
```

Comment:
I think I'd rather that this whole chunk be made more generic.  Like a function that takes a configuration name, a list of valid values, and a bool telling whether it's required, and then either inserts, or returns an error.

Reply:
I think we can push the code duplication reduction even farther, using the same approach described above in set_fields().